src/main/java/org/jenkinsci/plugins/p4/PerforceScmSource.java

```diff
@@ -0,0 +1,336 @@
+import hudson.model.TaskListener;
+import hudson.scm.RepositoryBrowser;
+import hudson.scm.RepositoryBrowsers;
+import hudson.scm.SCM;
+import hudson.util.FormValidation;
+import hudson.util.ListBoxModel;
+import jenkins.scm.api.SCMFile;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMHeadEvent;
+import jenkins.scm.api.SCMHeadObserver;
+import jenkins.scm.api.SCMProbe;
+import jenkins.scm.api.SCMProbeStat;
+import jenkins.scm.api.SCMRevision;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.SCMSourceCriteria;
+import jenkins.scm.api.SCMSourceDescriptor;
+import org.jenkinsci.plugins.p4.browsers.P4Browser;
+import org.jenkinsci.plugins.p4.client.ClientHelper;
+import org.jenkinsci.plugins.p4.client.ConnectionHelper;
+import org.jenkinsci.plugins.p4.credentials.P4CredentialsImpl;
+import org.jenkinsci.plugins.p4.populate.Populate;
+import org.jenkinsci.plugins.p4.workspace.StreamWorkspaceImpl;
+import org.jenkinsci.plugins.p4.workspace.Workspace;
+import org.jenkinsci.plugins.p4.workspace.WorkspaceDescriptor;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.logging.Logger;
+
+public class PerforceScmSource extends SCMSource {
+
+	private static Logger logger = Logger.getLogger(PerforceScmSource.class.getName());
+	private static String hack = "jenkins-master";
+
+	private final String credential;
+	private final String includes;
+	private final String charset;
+	private final String format;
+	private Populate populate;
+	private final P4Browser browser;
+
+	@DataBoundConstructor
+	public PerforceScmSource(String id, String credential, String includes, String charset, String format, P4Browser browser) {
+		super(id);
+
+		this.credential = credential;
+		this.includes = includes;
+		this.charset = charset;
+		this.format = format;
+		this.browser = browser;
+	}
+
+	@DataBoundSetter
+	public void setPopulate(Populate populate) {
+		this.populate = populate;
+	}
+
+	public String getCredential() {
+		return credential;
+	}
+
+	public String getIncludes() {
+		return includes;
+	}
+
+	public String getCharset() {
+		return charset;
+	}
+
+	public String getFormat() {
+		return format;
+	}
+
+	public P4Browser getBrowser() {
+		return browser;
+	}
+
+	public Populate getPopulate() {
+		return populate;
+	}
+
+	@Override
+	protected void retrieve(@CheckForNull SCMSourceCriteria criteria, @NonNull SCMHeadObserver observer, @CheckForNull SCMHeadEvent<?> event, @NonNull TaskListener listener) throws IOException, InterruptedException {
+		try {
+			List<PerforceHead> streams = getStreams(listener);
+			for (PerforceHead stream : streams) {
+				// check criteria; if null or not met, continue to next item
+				if (criteria == null) {
```

Comment:
We use `null` as a semi-special value because it can save instantiating the probe whereas a criteria that always returned `true` would require the probe be instantiated for each branch and that may be costly: for example instantiating the Git probe requires cloning the repository onto master while not instantiating the probe can identify all branches with just a single `git ls-remote`